### ISSUE TITLE
Add a Prometheus / Grafana interface

### ISSUE THREAD

**Issue by dirkwhoffmann** [dirkwhoffmann/vAmiga#850]
vAmiga features *activity monitors* that visualize the activity of certain components. 

<img width="868" alt="Bildschirmfoto 2024-12-22 um 18 10 22" src="https://github.com/user-attachments/assets/dcb0d815-7338-4362-89ae-5cb33c69f4e2" />

A more general approach would be to add a [Prometheus](https://prometheus.io) API and let [Grafana](https://grafana.com/oss/grafana/) do the visualization.

We could then have something like this running side by side with the emulator:

<img width="530" alt="Bildschirmfoto 2024-12-22 um 18 16 12" src="https://github.com/user-attachments/assets/4916878e-6190-4d60-894e-82cee34c0c09" />

**Comment by dirkwhoffmann** [comment]
I've been working on a proof-of-concept prototype to see what can be expected. 

Background: Prometheus pulls data from a data source via HTTP. As I don't want to add external dependencies to vAmiga, I integrated [httplib](https://github.com/yhirose/cpp-httplib) into vAmiga, which is header-only and sufficiently powerful for what we need. 

After that, I've implemented a tiny data provider that reports the current CPU load and the measured fps.

```C++
string
PromServer::generate_metrics()
{
    auto emuStats = emulator.getStats();

    string metrics =
    "# HELP cpu_load Host CPU load\n"
    "# TYPE cpu_load gauge\n"
    "cpu_load " + std::to_string(emuStats.cpuLoad) + "\n\n"

    "# HELP fps Frames per second\n"
    "# TYPE fps gauge\n"
    "fps " + std::to_string(emuStats.fps) + "\n\n";

    return metrics;
}
```

After adding vAmiga as a target, I've configured Prometheus to grab the data every second. This works well and can be verified in Prometheus' web API: 

<img width="1212" alt="Bildschirmfoto 2024-12-24 um 10 39 19" src="https://github.com/user-attachments/assets/72f0db9e-c1bc-4a1e-b3b7-01cd4e5484a0" />

Adding the Prometheus server as a data source to Grafana was also easy. Here is a Grafana panel displaying both metrics: 

<img width="603" alt="Bildschirmfoto 2024-12-24 um 10 39 58" src="https://github.com/user-attachments/assets/9512826f-d314-49ce-bfc0-77140f941677" />

Bottom line: Making vAmiga a Prometheus data provider is comparably easy and doable with reasonable effort.

**Comment by dirkwhoffmann** [comment]
Last update for today. The new PromServer has been integrated in the existing server framework as follows:

![classHierarchy](https://github.com/user-attachments/assets/ac0d63f6-6c26-4d19-9554-f969b925bf47)

All classes drawn in red are new.

**Comment by dirkwhoffmann** [comment]
OK, so this is how it works: 

To export metrics, start the Prometheus server in RetroShell: 

<img width="920" alt="Bildschirmfoto 2024-12-25 um 09 56 38" src="https://github.com/user-attachments/assets/f3cda62e-0901-4c3b-ad34-57d719e6dccc" />

The icon pointed to by the yellow arrow indicates that vAmiga is ready to provide data. The other icon (green arrow) appears when the first HTTP request comes in. 

By default, vAmiga provides metrics at `http://localhost:8082/metrics`. The `/metrics` endpoint is the one Prometheus uses by default to grab data from its targets. 

Directing the browser directly to the mentioned URL shows the currently exposed metrics:

```console
# TYPE vamiga_cpu_load gauge
vamiga_cpu_load{component="emulator"} 0.3945

# TYPE vamiga_fps gauge
vamiga_fps{component="emulator"} 60.0045

# TYPE vamiga_resyncs gauge
vamiga_resyncs{component="emulator"} 1

# TYPE vamiga_activity_copper gauge
vamiga_activity_copper{component="agnus"} 119.0000

# TYPE vamiga_activity_blitter gauge
vamiga_activity_blitter{component="agnus"} 0.0000

# TYPE vamiga_activity_disk gauge
vamiga_activity_disk{component="agnus"} 0.0000

# TYPE vamiga_activity_disk gauge
vamiga_activity_disk{component="agnus"} 0.0000

# TYPE vamiga_activity_audio gauge
vamiga_activity_audio{component="agnus"} 0.0000

# TYPE vamiga_activity_sprite gauge
vamiga_activity_sprite{component="agnus"} 0.0000

# TYPE vamiga_activity_bitplane gauge
vamiga_activity_bitplane{component="agnus"} 8000.0000

# TYPE vamiga_ciaa_idle_sec gauge
vamiga_ciaa_idle_sec{component="ciaa"} 13925

# TYPE vamiga_ciab_idle_sec gauge
vamiga_ciab_idle_sec{component="ciab"} 13925

# TYPE vamiga_cia_idle_sec_total gauge
vamiga_cia_idle_sec_total{component="ciaa"} 653940365

# TYPE vamiga_cia_idle_sec_total gauge
vamiga_cia_idle_sec_total{component="ciab"} 653940365

# TYPE vamiga_cia_idle_percentage gauge
vamiga_cia_idle_percentage{component="ciaa"} 0.9969

# TYPE vamiga_cia_idle_percentage gauge
vamiga_cia_idle_percentage{component="ciab"} 0.9969

# TYPE vamiga_mem_accesses gauge
vamiga_mem_accesses{component="memory",location="chip_ram",type="read"} 6.2903

# TYPE vamiga_mem_accesses gauge
vamiga_mem_accesses{component="memory",location="chip_ram",type="write"} 0.0000

# TYPE vamiga_mem_accesses gauge
vamiga_mem_accesses{component="memory",location="slow_ram",type="read"} 4318.2578

# TYPE vamiga_mem_accesses gauge
vamiga_mem_accesses{component="memory",location="slow_ram",type="write"} 2624.0211

# TYPE vamiga_mem_accesses gauge
vamiga_mem_accesses{component="memory",location="fast_ram",type="read"} 0.0000

# TYPE vamiga_mem_accesses gauge
vamiga_mem_accesses{component="memory",location="fast_ram",type="write"} 0.0000

# TYPE vamiga_mem_accesses gauge
vamiga_mem_accesses{component="memory",location="rom",type="read"} 9681.7736

# TYPE vamiga_mem_accesses gauge
vamiga_mem_accesses{component="memory",location="rom",type="write"} 0.0000

# TYPE vamiga_audio_buffer_exceptions gauge
vamiga_audio_buffer_exceptions{component="audio",type="overflow"} 0

# TYPE vamiga_audio_buffer_exceptions gauge
vamiga_audio_buffer_exceptions{component="audio",type="underflow"} 1

# TYPE vamiga_audio_samples gauge
vamiga_audio_samples{component="audio",type="consumed"} 1115306

# TYPE vamiga_audio_samples gauge
vamiga_audio_samples{component="audio",type="produced"} 0

# TYPE vamiga_audio_samples gauge
vamiga_audio_samples{component="audio",type="idle"} 1110414

# TYPE vamiga_audio_fill_level gauge
vamiga_audio_fill_level{component="audio"} 0.7639
```

Utilizing Grafana, the data collected by Prometheus can be visualized in various ways, e.g., like this:

<img width="912" alt="Bildschirmfoto 2024-12-25 um 09 58 01" src="https://github.com/user-attachments/assets/5f97a7e5-5962-4991-9c27-4c94232904c4" />

I am not in the mood to write a nice Grafana control panel myself right now, as the tool has many options, and I have never used it before. If somebody wants to do it, please go ahead and let me know which metrics would be cool to display (the currently captured metrics might not be the most meaningful). The infrastructure has been set up and fits nicely into the existing server architecture (SerServer, RshServer, GdbServer). I did this little project in the first place because some of my students used Prometheus and Grafana in their projects, and I was curious about how these tools work and how hard it is to incorporate them.

**Comment by dirkwhoffmann** [comment]
Supported in v3.0
